### ISSUE TITLE
Trust statistics seem off

### ISSUE THREAD

**Issue by Dmole** [Tribler/tribler#3657]
rm -r ~/.Tribler
apt install Tribler_7.1.0-exp2.deb
Tokens starts at 0 (seems like an easy way to game the system if there is any penalty for <0 )
Leave Tribler running for a bit with **0 Torrents**, then observe;
- Given = 1.2GB
- Taken = 0.4GB
- Tokens = 0.8GB

... think about it
- Should Taken =0?
- If not should Given=Taken?
- even if there were some explanation for this (like some chunks are discarded as they were provided by another node first) that would mean that actively trying to contribute could have a net negative impact on ones tokens, making the feature counter productive. 

Side note; please don't use the crazy date format (31-05-18), use `date --iso-8601` (2018-05-31); sorting largest to smallest units is the only rational format.

**Comment by devos50** [comment]
@Dmole that's quite interesting! When not downloading anonymously, taken should instead always be zero. The easiest way to debug this, is to explore the blocks on the Tribler blockchain I guess. Could you please provide your database with blocks? You can find this in your Tribler state directory, specifically in `~/.Tribler/sqlite/trustchain.db`. With this info, we might be able to figure out what happened 👍 (if you don't want to share this file publicly, you can also email it to me).

**Comment by Dmole** [comment]
One would hope with a fresh ~/.Tribler there would be nothing sensitive, but it's also kinda big so here is the tip;
```
du -sh ~/.Tribler/sqlite/trustchain.db | cut -f 1
17M

sqlite3 ~/.Tribler/sqlite/trustchain.db 'select count(*) from blocks;'
24442

sqlite3 ~/.Tribler/sqlite/trustchain.db 'select insert_time,sequence_number,link_sequence_number,tx from blocks limit 9;'
2018-05-31 21:31:55|17823|0|a4d4bdown7i733901310btotal_down12i1368913904318btotal_up13i15523933827842bup1i0
2018-05-31 21:31:55|1|17823|a4d4bdown1i010btotal_down1i08btotal_up7i73390132bup7i7339013
2018-05-31 21:31:55|1811|0|a4d4bdown9i13325219110btotal_down12i1362690080978btotal_up9i3689062682bup1i0
2018-05-31 21:31:55|2|0|a4d4bdown8i4441739710btotal_down8i444173978btotal_up7i73390132bup1i0
2018-05-31 21:31:55|3|1811|a4d4bdown1i010btotal_down8i444173978btotal_up9i1405912042bup9i133252191
2018-05-31 21:31:55|16100|2|a4d4bdown1i010btotal_down10i80832004978btotal_up13i13183900678272bup8i44417397
2018-05-31 21:31:55|1812|0|a4d4bdown9i13325718910btotal_down12i1364022652868btotal_up9i3689062682bup1i0
2018-05-31 21:31:55|4|1812|a4d4bdown1i010btotal_down8i444173978btotal_up9i2738483932bup9i133257189
2018-05-31 21:31:55|5|0|a4d4bdown8i4441906310btotal_down8i888364608btotal_up9i2738483932bup1i0
```
and I'll email you the rest.

**Comment by Thomasedv** [comment]
I can add that i also end up in negative from only having one Torrent in a seeding state. I tested this by removing the .Tribler folder, and then starting tribler and adding an already downloaded torrent, and let it seed for some time.

**Comment by devos50** [comment]
@Dmole @Thomasedv thank you for your help! I noticed this behaviour too when starting with a fresh Tribler instance. Next week I won't be at the university to work on it but I will try to investigate this as soon as I can when I'm back 👍

**Comment by devos50** [comment]
Related forum post: https://forum.tribler.org/t/uploads-and-downloads-are-miscalculated-tribler-7-1-0-exp2/4571

**Comment by Dmole** [comment]
2
History seems to disappear to, I built up a credit then downloaded something and only one day of data (no credit) was recorded...

3
v7.1.0-exp2 finds seeds faster but it's also quite crashy (I sent a few bug reports).

4
It seems impossible to have anonymity and -ve credits so Tribler should just not count anything below 0.

**Comment by devos50** [comment]
Thank you all for these reports! Fortunately, this should be relatively easy to debug since all blockchain records are public. This issue also makes it clear that we need a (very basic) blockchain explorer within the debug panel in order to debug it better.

We recently switched to a more sophisticated method of token payout so my first guess is that clients using the old method are communicating with clients that run a newer version. We have one more hard reset for the TrustChain structure in our pipeline (see #3572) so after that, this problem _might_ be resolved.

**Comment by Dmole** [comment]
What I said was a possibility

> actively trying to contribute could have a net negative impact on ones tokens, making the feature counter productive.

I have tested and confirmed to be a reality; since I first reported this I have just let Tribler run empty (0 torrents) and I got up to 1939 tokens then down to -10413 tokens...

**Comment by Dmole** [comment]
> It seems impossible to have anonymity and -ve credits so Tribler should just not count anything below 0.

The only way to make credits useful (unhackable) would be to make a history (of any sort) count for more than a weeks worth of seeding. (a fresh account should get less priority than a week old account with -1k points)

**Comment by qstokkink** [comment]
Overlap with #3731

**Comment by Dmole** [comment]
Technicaly this issue was first so not a duplicate.
Also there are 2 problems in this issue thread that are not mentioned in the other issue.
#tagMisuse

**Comment by devos50** [comment]
What actually goes over the write, seems to be way lower:

```
"statistics": {
                "bytes_down": 1701791165,
                "bytes_up": 128424579,
                "diff_time": 1517.9209060668945,
                "num_down": 1390178,
                "num_up": 1343046
            }
```

So, around 1.58GB goes over the wire but we are doing payouts of 6+ GB...

**Comment by devos50** [comment]
I adjusted the maximum number of circuits to be created for each download to one, which makes this issue easier to debug. It seems that that byte accounting of circuits is off, and this effect amplifies when we use more parallel circuits to download.

`byes_down` as reported by the circuit:

```
root@tribler-runner1:~# curl http://localhost:8085/debug/circuits | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   184  100   184    0     0  24210      0 --:--:-- --:--:-- --:--:-- 26285
{
    "circuits": [
        {
            "bytes_down": 266833732,
            "bytes_up": 516,
            "created": 1538485831.260635,
            "goal_hops": 1,
            "hops": [
                {
                    "host": "37.48.77.232:35020"
                }
            ],
            "id": 3095613092,
            "state": "CLOSING"
        }
    ]
}
```

Actual data that went over the wire:

```
            "statistics": {
                "bytes_down": 139485921,
                "bytes_up": 11585823,
                "diff_time": 118.32790017127991,
                "num_down": 124441,
                "num_up": 119183
            }
```

**Comment by devos50** [comment]
I fixed a part of the issue in https://github.com/Tribler/py-ipv8/pull/315, but it seems that there is another bug where a single payout is sent twice to the other party. I will investigate this too.

**Comment by devos50** [comment]
Payouts are now accurate when using a single circuit to download a 1.5GB file:

Over the wire:
```
"statistics": {
                "bytes_down": 1690552461,
                "bytes_up": 132366281,
                "diff_time": 1942.5653347969055,
                "num_down": 1405838,
                "num_up": 1390456
            }
```

And we did a payout of 1.57GB 👍

**Comment by devos50** [comment]
Also confirmed to work when downloading with multiple hops 👍

**Comment by Dmole** [comment]
I turned on credit mining but lost ~10GB of  ratio/trust.
this bug should not be closed IMO.

**Comment by synctext** [comment]
@Dmole
Sorry for losing that 10GB of credit. Correct I think, it's the initial download that is causing this. It should be smarter then that. Its overinvesting. We had 4 people on credit mining in past 5-ish years. 
Proven very difficult to get a token economy going smoothly. Please bare with us..

**Comment by devos50** [comment]
Yes, it is most likely the initial download which is causing this. The credit mining algorithm first downloads the whole file and then seeds it to others. It's a naive but basic algorithm. We will definitely improve the credit mining algorithm during the coming releases.

This issue however was meant for the double accounting bug when doing a regular download, which is fixed now.

**Comment by Dmole** [comment]
I made this bug report with the idea of getting statistics reasonable, but if you want to re-classify it as just a doubling bug I can make new bug reports for the other sub issues if that helps.

A token economy would be more likely to happen after the statistics and traid are reasonable (not yet).

**Comment by devos50** [comment]
If you notice that something is off, please create separate issues for it. This issue should have been made a bit more specific by us (i.e. by modifying the title to make it more specific), so we will keep that in mind.

> A token economy would be more likely to happen after the statistics and traid are reasonable (not yet).

Building an operational token economy requires many iterations and testing. Now that downloading seems to work fine, fixing the token mining algorithm is becoming a more important issue. Also, explaining the process to the end-user in a clear manner is challenging, but we'll get there based on your feedback and the feedback of others 👍 

Also, the market requires more polish, especially the user-interface part. Last months, I spent much time researching and implementing the decentralized algorithm for trading (and wrote an article about it). Getting it fully operational is the next step.

**Comment by Dmole** [comment]
5 years later and the Trust statistics are still untrustworthy, but at least hidden at the moment;

> v7.12.1
> The token balance is now hidden until the algorithm for calculating the balance is updated.
